README.md
# keymaker-toolbox

Toolbox of keymaker.

[![Travis build status](http://img.shields.io/travis/hemslo/keymaker-toolbox.svg?style=flat)](https://travis-ci.org/hemslo/keymaker-toolbox)
[![Code Climate](https://codeclimate.com/github/hemslo/keymaker-toolbox/badges/gpa.svg)](https://codeclimate.com/github/hemslo/keymaker-toolbox)
[![Test Coverage](https://codeclimate.com/github/hemslo/keymaker-toolbox/badges/coverage.svg)](https://codeclimate.com/github/hemslo/keymaker-toolbox)
[![Dependency Status](https://david-dm.org/hemslo/keymaker-toolbox.svg)](https://david-dm.org/hemslo/keymaker-toolbox)
[![devDependency Status](https://david-dm.org/hemslo/keymaker-toolbox/dev-status.svg)](https://david-dm.org/hemslo/keymaker-toolbox#info=devDependencies)

### How it works

Use PBKDF2 with SHA-256 to derive a key, then replace every byte with supplied character sets(lower, upper, number, special).


CHANGELOG.md
### [0.0.1](https://github.com/hemslo/keymaker-toolbox/releases/tag/v0.0.1)

- The first release

### [0.0.2](https://github.com/hemslo/keymaker-toolbox/releases/tag/v0.0.2)

- Upgrade to babel 6

### [0.0.3](https://github.com/hemslo/keymaker-toolbox/releases/tag/v0.0.3)

- Move sjcl to externals


test/unit/keymaker-toolbox.js
import keymakerToolbox from '../../src/keymaker-toolbox';

describe('keymakerToolbox', () => {
  describe('makeKey function', () => {
    it('should return a key', () => {
      expect(keymakerToolbox.makeKey('password', 'salt')).to.equal('$s2d$0TF2c6`k$2e$9PqySg`0PVs:BRJ');
    });

    it('should return hex when all false', () => {
      expect(keymakerToolbox.makeKey('password', 'salt', 100000, 32, false, false, false, false)).to.equal('0394a2ede332c9a13eb82e9b24631604c31df978b4e2f0fbd2c549944f9d79a5');
    });
  });
});


gulpfile.babel.js
import gulp  from 'gulp';
import loadPlugins from 'gulp-load-plugins';
import del  from 'del';
import glob  from 'glob';
import path  from 'path';
import {Instrumenter} from 'isparta';
import webpack from 'webpack';
import webpackStream from 'webpack-stream';
import source  from 'vinyl-source-stream';

import mochaGlobals from './test/setup/.globals';
import manifest  from './package.json';

// Load all of our Gulp plugins
const $ = loadPlugins();

// Gather the library data from `package.json`
const config = manifest.babelBoilerplateOptions;
const mainFile = manifest.main;
const destinationFolder = path.dirname(mainFile);
const exportFileName = path.basename(mainFile, path.extname(mainFile));

function cleanDist(done) {
  del([destinationFolder]).then(() => done());
}

function cleanTmp(done) {
  del(['tmp']).then(() => done());
}

function onError() {
  $.util.beep();
}

// Lint a set of files
function lint(files) {
  return gulp.src(files)
    .pipe($.plumber())
    .pipe($.eslint())
    .pipe($.eslint.format())
    .pipe($.eslint.failOnError())
    .pipe($.jscs())
    .pipe($.jscs.reporter('fail'))
    .on('error', onError);
}

function lintSrc() {
  return lint('src/**/*.js');
}

function lintTest() {
  return lint('test/**/*.js');
}

function lintGulpfile() {
  return lint('gulpfile.babel.js');
}

function build() {
  return gulp.src(path.join('src', config.entryFileName + '.js'))
    .pipe($.plumber())
    .pipe(webpackStream({
      output: {
        filename: exportFileName + '.js',
        libraryTarget: 'umd',
        library: config.mainVarName
      },
      module: {
        loaders: [
          { test: /\.js$/, exclude: /node_modules/, loader: 'babel-loader' }
        ]
      },
      externals: {
        'sjcl': 'sjcl'
      },
      devtool: 'source-map'
    }))
    .pipe(gulp.dest(destinationFolder))
    .pipe($.filter(['*', '!**/*.js.map']))
    .pipe($.rename(exportFileName + '.min.js'))
    .pipe($.sourcemaps.init({ loadMaps: true }))
    .pipe($.uglify())
    .pipe($.sourcemaps.write('./'))
    .pipe(gulp.dest(destinationFolder));
}

function _mocha() {
  return gulp.src(['test/setup/node.js', 'test/unit/**/*.js'], {read: false})
    .pipe($.mocha({
      reporter: 'dot',
      globals: Object.keys(mochaGlobals.globals),
      ignoreLeaks: false
    }));
}

function _registerBabel() {
  require('babel-register');
}

function test() {
  _registerBabel();
  return _mocha();
}

function coverage(done) {
  _registerBabel();
  gulp.src(['src/**/*.js'])
    .pipe($.istanbul({ instrumenter: Instrumenter }))
    .pipe($.istanbul.hookRequire())
    .on('finish', () => {
      return test()
        .pipe($.istanbul.writeReports())
        .on('end', done);
    });
}

const watchFiles = ['src/**/*', 'test/**/*', 'package.json', '**/.eslintrc', '.jscsrc'];

// Run the headless unit tests as you make changes.
function watch() {
  gulp.watch(watchFiles, ['test']);
}

function testBrowser() {
  // Our testing bundle is made up of our unit tests, which
  // should individually load up pieces of our application.
  // We also include the browser setup file.
  const testFiles = glob.sync('./test/unit/**/*.js');
  const allFiles = ['./test/setup/browser.js'].concat(testFiles);

  // Lets us differentiate between the first build and subsequent builds
  var firstBuild = true;

  // This empty stream might seem like a hack, but we need to specify all of our files through
  // the `entry` option of webpack. Otherwise, it ignores whatever file(s) are placed in here.
  return gulp.src('')
    .pipe($.plumber())
    .pipe(webpackStream({
      watch: true,
      entry: allFiles,
      output: {
        filename: '__spec-build.js'
      },
      module: {
        loaders: [
          // This is what allows us to author in future JavaScript
          { test: /\.js$/, exclude: /node_modules/, loader: 'babel-loader' },
          // This allows the test setup scripts to load `package.json`
          { test: /\.json$/, exclude: /node_modules/, loader: 'json-loader' }
        ]
      },
      plugins: [
        // By default, webpack does `n=>n` compilation with entry files. This concatenates
        // them into a single chunk.
        new webpack.optimize.LimitChunkCountPlugin({ maxChunks: 1 })
      ],
      devtool: 'inline-source-map'
    }, null, function() {
      if (firstBuild) {
        $.livereload.listen({port: 35729, host: 'localhost', start: true});
        var watcher = gulp.watch(watchFiles, ['lint']);
      } else {
        $.livereload.reload('./tmp/__spec-build.js');
      }
      firstBuild = false;
    }))
    .pipe(gulp.dest('./tmp'));
}

// Remove the built files
gulp.task('clean', cleanDist);

// Remove our temporary files
gulp.task('clean-tmp', cleanTmp);

// Lint our source code
gulp.task('lint-src', lintSrc);

// Lint our test code
gulp.task('lint-test', lintTest);

// Lint this file
gulp.task('lint-gulpfile', lintGulpfile);

// Lint everything
gulp.task('lint', ['lint-src', 'lint-test', 'lint-gulpfile']);

// Build two versions of the library
gulp.task('build', ['lint', 'clean'], build);

// Lint and run our tests
gulp.task('test', ['lint'], test);

// Set up coverage and run tests
gulp.task('coverage', ['lint'], coverage);

// Set up a livereload environment for our spec runner `test/runner.html`
gulp.task('test-browser', ['lint', 'clean-tmp'], testBrowser);

// Run the headless unit tests as you make changes.
gulp.task('watch', watch);

// An alias of test
gulp.task('default', ['test']);


src/keymaker-toolbox.js
import sjcl from 'sjcl';

const Characters = {
  lower: 'abcdefghijklmnopqrstuvwxyz'.split(''),
  upper: 'ABCDEFGHIJKLMNOPQRSTUVWXYZ'.split(''),
  number: '0123456789'.split(''),
  special: '!"#$%&\'()*+,-./:;<=>?@[\\]^_`{|}~'.split('')
};

const keymakerToolbox = {
  makeKey(password, salt, iterations=100000, keylen=32, lower=true, upper=true, number=true, special=true) {
    let dk = sjcl.misc.pbkdf2(password,
                              sjcl.codec.utf8String.toBits(salt),
                              iterations,
                              keylen * 8);
    let hex = sjcl.codec.hex.fromBits(dk);

    let types = [];
    const allTypes = ['lower', 'upper', 'number', 'special'];
    [lower, upper, number, special].forEach((e, i) => {
      if (e) {
        types.push(allTypes[i]);
      }
    });

    if (types.length === 0 || keylen < types.length) {
      return hex;
    }

    let decs = hex.match(/.{2}/g).map(i => parseInt(i, 16));
    let characters = decs.map(dec => {
      let type = types[dec % types.length];
      return Characters[type][dec % Characters[type].length];
    });

    // ensure all types at least one
    let indices = characters.map((e, i) => i);
    let index = 0;
    types.forEach(type => {
      let i = decs[index] % indices.length;
      index = indices[i];
      indices.splice(i, 1);
      characters[index] = Characters[type][decs[index] % Characters[type].length];
    });
    return characters.join('');
  }
};

export default keymakerToolbox;
